python/cugraph/cugraph/gnn/data_loading/dist_sampler.py

```diff
@@ -0,0 +1,349 @@
+                    "This might cause your training loop to hang due to uneven inputs."
+                )
+
+            batch_id_start = t.cumsum(dim=0)[rank] - t[0]
+        else:
+            batch_id_start = 0
+
+        for i, current_seeds in enumerate(nodes):
+            current_batches = torch.arange(
+                batch_id_start + i * batches_per_call,
+                batch_id_start + (i + 1) * batches_per_call,
+                device="cuda",
+                dtype=torch.int32,
+            )
+
+            current_batches = current_batches.repeat_interleave(batch_size)[
+                : len(current_seeds)
+            ]
+
+            minibatch_dict = self.sample_batches(
+                seeds=current_seeds,
+                batch_ids=current_batches,
+                random_state=random_state,
+            )
+            self.__writer.write_minibatches(minibatch_dict)
+
+    @property
+    def is_multi_gpu(self):
+        return isinstance(self.__graph, pylibcugraph.MGGraph)
+
+    @property
+    def _local_seeds_per_call(self):
+        return self.__local_seeds_per_call
+
+    @property
+    def _graph(self):
+        return self.__graph
+
+
+class UniformNeighborSampler(DistSampler):
+    def __init__(
+        self,
+        graph: Union[pylibcugraph.SGGraph, pylibcugraph.MGGraph],
+        writer: DistSampleWriter,
+        *,
+        local_seeds_per_call: int = 32768,
+        fanout: List[int] = [-1],
+        prior_sources_behavior: str = "exclude",
+        deduplicate_sources: bool = True,
+        compression: str = "COO",
+        compress_per_hop: bool = False,
+        with_replacement: bool = False,
+    ):
+        super().__init__(graph, writer, local_seeds_per_call=local_seeds_per_call)
+        self.__fanout = fanout
+        self.__prior_sources_behavior = prior_sources_behavior
+        self.__deduplicate_sources = deduplicate_sources
+        self.__compress_per_hop = compress_per_hop
+        self.__compression = compression
+        self.__with_replacement = with_replacement
+
+    def sample_batches(
+        self, seeds: TensorType, batch_ids: TensorType, random_state: int = 0
+    ):
+        # FIXME allow skipping of the synchronization logic with a boolean
+        # flag that assumes all ranks have the same number of batches.
+        if self.is_multi_gpu:
+            rank = torch.distributed.get_rank()
+            handle = pylibcugraph.ResourceHandle(
+                cugraph_comms_get_raft_handle().getHandle()
+            )
+
+            local_label_list = torch.unique(batch_ids)
+            local_label_to_output_comm_rank = torch.full(
+                (len(local_label_list),), rank, dtype=torch.int32, device="cuda"
+            )
+
+            num_batches = torch.tensor(
+                [len(local_label_list)], device="cuda", dtype=torch.int64
+            )
+            torch.distributed.all_reduce(num_batches, op=torch.distributed.ReduceOp.SUM)
```

Comment:
Strongly suggest making functions for gatherig num_batches, label_list etc

Reply:
Agreed, and I need to add the flag to skip this logic.